ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,199 @@
+    this.metadata.put(queryId, binPackedPersistentQueryMetadata);
+    this.listeners.put(queryId, binPackedPersistentQueryMetadata.getListener());
+  }
+
+  protected StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+                                                                                         ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (Map.Entry<QueryId, QueryMetadata.Listener> listener: listeners.entrySet()) {
+        listener.getValue().onError(metadata.get(listener.getKey()), queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+      );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.allMetadata();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.localThreadsMetadata()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void addToplogy(final Topology topology) {
+    queryMap.put(topology.getName(), topology);
+    kafkaStreams.addTopology(topology);
+  }
+
+  public void stop(final QueryId queryId) {
+    queryMap.remove(queryId.toString());
+    metadata.remove(queryId);
+    listeners.remove(queryId);
+    kafkaStreams.removeTopology(queryId.toString());
+  }
+
+  public void restart(final QueryId queryId) {
+    final KafkaStreams newKafkaStreams = kafkaStreamsBuilder.build(null, streamsProperties);
+    for (Topology topology: queryMap.values()) {
+      newKafkaStreams.addTopology(topology);
+    }
+    newKafkaStreams.start();
```

Comment:
would it make sense to reverse the order here? especially if someone has static membership configured

Reply:
I may not have added it in the initial skeleton API, but it's in the Pt. 1 PR. I'm hoping we can have that merged soon. But yeah, for restarted queries we should use the collection-based API rather than adding them one-at-a-time